nncf/common/pruning/node_selector.py

```diff
@@ -183,79 +191,113 @@ def _get_multiforward_nodes(self, graph: NNCFGraph) -> List[List[NNCFNode]]:
+        pruning algo have a correct correspondent closing node on each path form self to outputs.
+
+        :param graph: Graph to work with.
+        :param can_prune_after_check: Dict of node indices vs the decision made by previous steps;
+            the decision is true only for the nodes that do not conflict with mask propagation and
+            are supported by the NNCF pruning algorithm
+        :return: Pruning node analysis after model analyzer, pruning algo compatibility and pruning dimensions checks.
+        """
+        mask_prop_algo = MaskPropagationAlgorithm(graph, self._pruning_operator_metatypes)
+        can_prune_by_dim = mask_prop_algo.symbolic_mask_propagation(self._prune_operations_types, can_prune_after_check)
+
+        can_prune_for_prunable_layers = \
+            {node_id: can_prune_after_check[node_id].join(can_prune_by_dim[node_id]) for node_id in can_prune_by_dim}
+
+        can_prune_updated = can_prune_after_check.copy()
+        can_prune_updated.update(can_prune_for_prunable_layers)
+        return can_prune_updated
+
+    def _should_prune_groups_analysis(self, graph: NNCFGraph, pruned_nodes_clusterization: Clusterization,
+                                      can_prune: Dict[int, PruningAnalysisDecision]) \
+            -> Dict[int, PruningAnalysisDecision]:
         """
         Check whether all nodes in group can be pruned based on user-defined constraints and
-        connections inside the network. Otherwise the whole group cannot be pruned.
+        model analysis. Otherwise the whole group cannot be pruned.
 
         :param graph: Graph to work with.
+        :param pruned_nodes_clusterization: Clusterization with pruning nodes groups.
+        :param can_prune: Complete pruning analysis about each graph node.
+        :return:
+        """
+        should_prune = {}
+        for cluster in pruned_nodes_clusterization.get_all_clusters():
+            should_prune_decisions = [self._is_module_prunable(graph, node) for node in cluster.elements]
+            can_prune_decisions = [can_prune[node.node_id] for node in cluster.elements]
+            decisions = [can.join(should) for can, should in zip(can_prune_decisions, should_prune_decisions)]
+            if not all(decisions):
+                updated_decisions = {}
+                for node, decision in zip(cluster.elements, decisions):
+                    if decision:
+                        updated_decisions[node.node_id] = \
+                            PruningAnalysisDecision(False, PruningAnalysisReason.IN_GROUP_OF_UNPRUNABLE)
+                    else:
+                        updated_decisions[node.node_id] = decision
+
+                should_prune.update(updated_decisions)
+
+        can_prune_updated = can_prune.copy()
+        can_prune_updated.update(should_prune)
+        return can_prune_updated
+
+    def _filter_pruning_groups(self, pruned_nodes_clusterization: Clusterization,
```

Comment:
You are not right, see line 273

Reply:
Still `filter` as a verb can be confused with `filter pruning` as an algo name. And the description doesn't mention that it will remove unprunable groups.

Reply:
Done, please check